src/semantic-highlighting.ts

```diff
@@ -0,0 +1,168 @@
+// Mirrors the structure in the semantic highlighting proposal for LSP.
+interface SemanticHighlightingParams {
+  // The text document that has to be decorated with the semantic highlighting
+  // information.
+  textDocument: VersionedTextDocumentIdentifier;
+  // An array of semantic highlighting information.
+  lines: SemanticHighlightingInformation[];
+}
+
+// A SemanticHighlightingToken decoded from the base64 data sent by clangd.
+interface SemanticHighlightingToken {
+  // Start column for this token.
+  character: number;
+  // Length of the token.
+  length: number;
+  // The TextMate scope index to the clangd scope lookup table.
+  scopeIndex: number;
+  kind: string;
+}
+
+// A line of decoded highlightings from the data clangd sent.
+export interface SemanticHighlightingLine {
+  // The zero-based line position in the text document.
+  line: number;
+  // All SemanticHighlightingTokens on the line.
+  tokens: SemanticHighlightingToken[];
+}
+
+export class SemanticHighlightingFeature implements StaticFeature {
+  private scopeTable: string[][] = [];
+
+  constructor(client: BaseLanguageClient, context: ExtensionContext) {
+    context.subscriptions.push(
+      client.onDidChangeState(({ newState }) => {
+        if (newState === State.Running) {
+          const notification = new NotificationType<SemanticHighlightingParams, void>('textDocument/semanticHighlighting');
+          client.onNotification(notification, this.handleNotification.bind(this));
+        }
+      })
+    );
+  }
+
+  initialize(capabilities: ServerCapabilities) {
+    const serverCapabilities: ServerCapabilities & { semanticHighlighting?: { scopes: string[][] } } = capabilities;
+    if (!serverCapabilities.semanticHighlighting) return;
+
+    this.scopeTable = serverCapabilities.semanticHighlighting.scopes;
+  }
+
+  fillClientCapabilities(capabilities: ClientCapabilities) {
+    const textDocumentCapabilities: TextDocumentClientCapabilities & { semanticHighlightingCapabilities?: { semanticHighlighting: boolean } } = capabilities.textDocument!;
+    textDocumentCapabilities.semanticHighlightingCapabilities = {
+      semanticHighlighting: true,
+    };
+  }
+
+  async handleNotification(params: SemanticHighlightingParams) {
+    // use https://github.com/jackguo380/vim-lsp-cxx-highlight to do highlighting
+    const lines: SemanticHighlightingLine[] = params.lines.map((line) => ({ line: line.line, tokens: this.decodeTokens(line.tokens) }));
+    const symbols: any[] = [];
+    const skipped: any[] = [];
+
+    for (const line of lines) {
+      for (const token of line.tokens) {
+        symbols.push({
+          id: 0,
+          kind: token.kind,
+          ranges: [Range.create(line.line, token.character, line.line, token.character + token.length)],
+          parentKind: 'Unknown', // FIXME
+          storage: 'None', // FIXME
+        });
+        if (token.kind === 'InactiveCode') {
```

Comment:
if the kind is InactiveCode, then you don't want to do the symbols.push above - one or the other